marshal.nim can't be compiled for the cpp backend

try to run `nim cpp --nimcache:/tmp lib/pure/marshal.nim` in your Nim base directory. It will produce something like:
```
Error: execution of an external compiler program 'g++ -c  -w -w -fpermissive  -I/home/skilchen/nim/nim/lib -o /tmp/stdlib_marshal.cpp.o /tmp/stdlib_marshal.cpp' failed with exit code: 1

/tmp/stdlib_marshal.cpp: In function ‘void to_aWmVqxX2DsT7MwOicWdS5A(NimStringDesc*, NU8*)’:
/tmp/stdlib_marshal.cpp:2541:124: error: invalid initialization of reference of type ‘NU8 (&)[32] {aka unsigned char (&)[32]}’ from expression of type ‘NU8* {aka unsigned char*}’
  loadAny_VX2RGdenDcgA5ChmFXKVyw(newStringStream_0dQ0A0pna0dG9cMxo4TbLKQ(data), toAny_Qiaji9cqG669bj1Pe0K0PSHwmarshal(Result), tab);
                                                                                                                            ^
In file included from /tmp/stdlib_marshal.cpp:10:0:
/tmp/stdlib_marshal.cpp:2499:54: note: in passing argument 1 of ‘tyObject_Any_bEXZHc2yZOebOeouZDUZ7A toAny_Qiaji9cqG669bj1Pe0K0PSHwmarshal(NU8 (&)[32])’
 static N_INLINE(tyObject_Any_bEXZHc2yZOebOeouZDUZ7A, toAny_Qiaji9cqG669bj1Pe0K0PSHwmarshal)(tySet_tyChar_nmiMWKVIe46vacnhAFrQvw& x) {
                                                      ^
/home/skilchen/nim/nim/lib/nimbase.h:95:50: note: in definition of macro ‘N_INLINE’
 #  define N_INLINE(rettype, name) inline rettype name
```

--- a/compiler/ccgexprs.nim
+++ b/compiler/ccgexprs.nim
@@ -1596,10 +1596,11 @@ proc genSwap(p: BProc, e: PNode, d: var TLoc) =
   genAssignment(p, a, b, {})
   genAssignment(p, b, tmp, {})
 
-proc rdSetElemLoc(conf: ConfigRef; a: TLoc, setType: PType): Rope =
+proc rdSetElemLoc(conf: ConfigRef; a: TLoc, typ: PType): Rope =
   # read a location of an set element; it may need a subtraction operation
   # before the set operation
   result = rdCharLoc(a)
+  let setType = typ.skipTypes(abstractPtrs)
   assert(setType.kind == tySet)
   if firstOrd(conf, setType) != 0:
     result = "($1- $2)" % [result, rope(firstOrd(conf, setType))]

--- a/compiler/ccgtypes.nim
+++ b/compiler/ccgtypes.nim
@@ -162,9 +162,9 @@ proc mapType(conf: ConfigRef; typ: PType): TCTypeKind =
     var base = skipTypes(typ.lastSon, typedescInst)
     case base.kind
     of tyOpenArray, tyArray, tyVarargs: result = ctPtrToArray
-    #of tySet:
-    #  if mapSetType(base) == ctArray: result = ctPtrToArray
-    #  else: result = ctPtr
+    of tySet:
+      if mapSetType(conf, base) == ctArray: result = ctPtrToArray
+      else: result = ctPtr
     # XXX for some reason this breaks the pegs module
     else: result = ctPtr
   of tyPointer: result = ctPtr
@@ -641,10 +641,11 @@ proc getTypeDescAux(m: BModule, origTyp: PType, check: var IntSet): Rope =
                     compileToCpp(m): "&" else: "*"
     var et = origTyp.skipTypes(abstractInst).lastSon
     var etB = et.skipTypes(abstractInst)
-    if etB.kind in {tyArray, tyOpenArray, tyVarargs}:
-      # this is correct! sets have no proper base type, so we treat
-      # ``var set[char]`` in `getParamTypeDesc`
-      et = elemType(etB)
+    if mapType(m.config, t) == ctPtrToArray:
+      if etB.kind == tySet:
+        et = getSysType(m.g.graph, unknownLineInfo(), tyUInt8)
+      else:
+        et = elemType(etB)
       etB = et.skipTypes(abstractInst)
       star[0] = '*'
     case etB.kind

--- a/tests/ccgbugs/t8967.nim
+++ b/tests/ccgbugs/t8967.nim
@@ -1,0 +1,10 @@
+discard """
+  targets: "c cpp"
+"""
+
+import marshal
+
+let orig: set[char] = {'A'..'Z'}
+let m = $$orig
+let old = to[set[char]](m)
+doAssert orig - old == {}

